db/system_distributed_keyspace.cc

```diff
@@ -329,41 +352,73 @@ system_distributed_keyspace::expire_cdc_topology_description(
             false).discard_result();
 }
 
-static set_type_impl::native_type prepare_cdc_streams(const std::vector<cdc::stream_id>& streams) {
-    set_type_impl::native_type ret;
-    for (auto& s: streams) {
-        ret.push_back(data_value(s.to_bytes()));
-    }
-    return ret;
-}
-
-// cdc_streams_descriptions mutation
-static future<mutation> get_mutation(
+// "cdc_streams_descriptions_v2" mutation
+static future<std::vector<mutation>> get_mutations(
         const database& db,
         db_clock::time_point time,
-        const std::vector<cdc::stream_id>& streams) {
+        const cdc::topology_description& desc) {
     auto s = db.find_schema(system_distributed_keyspace::NAME, system_distributed_keyspace::CDC_DESC);
-    mutation m(s, partition_key::from_singular(*s, time));
-    m.set_cell(clustering_key::make_empty(), to_bytes("streams"),
-            make_set_value(cdc_streams_set_type, prepare_cdc_streams(streams)), api::new_timestamp());
-    co_return m;
+
+    auto ts = api::new_timestamp();
+    std::vector<mutation> res;
+    res.emplace_back(s, partition_key::from_singular(*s, time));
+    size_t size_estimate = 0;
+    for (auto& e : desc.entries()) {
+        // We want to keep each mutation below ~1 MB.
+        if (size_estimate >= 1000 * 1000) {
+            res.emplace_back(s, partition_key::from_singular(*s, time));
+            size_estimate = 0;
+        }
+
+        set_type_impl::native_type streams;
+        streams.reserve(e.streams.size());
+        for (auto& stream : e.streams) {
+            streams.push_back(data_value(stream.to_bytes()));
+        }
+
+        // We estimate 20 bytes per stream ID.
+        // Stream IDs themselves weigh 16 bytes each (2 * sizeof(int64_t))
+        // but there's metadata to be taken into account.
+        // It has been verified experimentally that 20 bytes per stream ID is a good estimate.
+        size_estimate += e.streams.size() * 20;
+        res.back().set_cell(clustering_key::from_singular(*s, dht::token::to_int64(e.token_range_end)),
+                to_bytes("streams"), make_set_value(cdc_streams_set_type, std::move(streams)), ts);
+
+        co_await make_ready_future<>(); // maybe yield
+    }
+
+    co_return res;
 }
 
 future<>
 system_distributed_keyspace::create_cdc_desc(
         db_clock::time_point time,
-        const std::vector<cdc::stream_id>& streams,
+        const cdc::topology_description& desc,
         context ctx) {
     using namespace std::chrono_literals;
-    auto m = co_await get_mutation(_qp.db(), time, streams);
-    co_return co_await _sp.mutate(
+
+    auto ms = co_await get_mutations(_qp.db(), time, desc);
+    // The number of mutations will be around 30 in a 100-node cluster with 64 shards per node,
+    // so in practice limiting concurrency here is not necessary, but just to be safe...
+    co_await max_concurrent_for_each(ms, 100, [&] (mutation& m) -> future<> {
```

Comment:
This requires 100MB just for this operation. It can stress a smaller shard.

I'd limit it to 20MB.